pkg/ingester/activeseries/active_series.go

```diff
@@ -84,50 +88,83 @@ type seriesEntry struct {
 	deleted bool // This series was marked as deleted, so before purging we need to remove the refence to it from the deletedSeries.
 }
 
-func NewActiveSeries(asm *asmodel.Matchers, timeout time.Duration) *ActiveSeries {
-	c := &ActiveSeries{matchers: asm, timeout: timeout}
+func NewActiveSeries(
+	asm *asmodel.Matchers,
+	timeout time.Duration,
+	cat costattribution.Tracker,
+) *ActiveSeries {
+	c := &ActiveSeries{
+		matchers: asm, timeout: timeout, cat: cat,
+	}
 
 	// Stripes are pre-allocated so that we only read on them and no lock is required.
 	for i := 0; i < numStripes; i++ {
-		c.stripes[i].reinitialize(asm, &c.deleted)
+		c.stripes[i].reinitialize(asm, &c.deleted, cat)
 	}
 
 	return c
 }
 
 func (c *ActiveSeries) CurrentMatcherNames() []string {
-	c.matchersMutex.RLock()
-	defer c.matchersMutex.RUnlock()
+	c.configMutex.RLock()
+	defer c.configMutex.RUnlock()
 	return c.matchers.MatcherNames()
 }
 
+// Function to compare two Tracker instances
+func areTrackersEqual(t1, t2 costattribution.Tracker) bool {
+	if t1 == t2 {
+		// If both trackers are the same pointer (including nil), they are equal
+		return true
+	}
+
+	// Use type assertion to check if both are NoopTracker
+	_, isNoop1 := t1.(*costattribution.NoopTracker)
+	_, isNoop2 := t2.(*costattribution.NoopTracker)
+
+	// If both are NoopTracker instances, treat them as equal
+	return isNoop1 && isNoop2
+}
+
+func (c *ActiveSeries) ConfigDiffers(ctCfg asmodel.CustomTrackersConfig, caCfg costattribution.Tracker) bool {
+	if ctCfg.String() != c.CurrentConfig().String() {
+		return true
+	}
+	return !areTrackersEqual(caCfg, c.CurrentCostAttributionTracker())
+}
```

Comment:
Since we have `ConfigDiffers` now, we don't need `CurrentConfig` or `CurrentCostAttributionTracker` methods, and we don't need to take mutex twice for them.

Reply:
addressed in this commit https://github.com/grafana/mimir/pull/9733/commits/698a5c6b682f235a88584e6b55729419ea394a56#diff-74600e3203a9faaa94f7231d161362da3e3223aa024eb324b903e5fc27bb9f3dR116